### PR TITLE
Restructure ozone luna 

### PR DESCRIPTION
**Pull request by sunnivin**
### Description of changes
Generalization of the scientific ozone-luna exploration done by @ziu1986 described in #1232. New calling structure implemented as suggested by @billsacks  in #1276. 

### Specific notes

Contributors other than yourself, if any: @ziu1986, @danicalombardozzi, @billsacks 

CTSM Issues Fixed (include github issue #): #1232 

Are answers expected to change (and if so in what way)? 
Yes, the history fields TLAI, GPP, NPP, carbon/nitrogen content, stomatal conductance, and photosynthesis are expected to drop by 10-20 % in accordance with earlier works. 

Any User Interface Changes (namelist or namelist defaults changes)? 
Yes, In the file `bld/namelist_files/namelist_definition_ctsm.xml` the namelist has been updated: `use_ozone --> ozone_method`. Valid options: `ozone_method= unset,stress_lombardozzi2015,stress_falk`. Furthermore the namelist variable `o3_ppbv` has also been added. 

Testing performed, if any: 
The new implementation has reproduced the scientific results obtained in #1232 for the single point in Brazil `--resolution 1x1_brazil` whit the `--compset 2000_DATM%GSWP3v1_CLM50%BGC_SICE_SOCN_SROF_SGLC_SWAV`.

### CODE REVIEW

**Comment by sunnivin** [comment]
Still not sure if the namelist variable `ozone_method= unset, stress_lombardozzi2015, stress_falk` has the optimal name.

File: src/biogeophys/OzoneMod.F90
Comment (billsacks) [inline src/biogeophys/OzoneMod.F90:671]: - [ ] [OPTIONAL] Since the coefficients set here are only needed at the end of the day, it might make sense to put an end-of-day conditional here, so this routine returns without doing anything on most time steps rather than doing unnecessary work. I'm a little torn on this, because I generally don't like having logic in one module that depends on logic in another module like this, but I'd lean towards adding this. (A better solution, which could make this dependency more obvious, could be to introduce a new function like is_time_to_run_luna, and replace the current checks of is_end_curr_day with a call to that function. This function would merely wrap is_end_curr_day, so the behavior would be the same, but the naming would give a clear signal to the code reader / maintainer about the reason we're doing this check – and then a change in the logic for when to run luna would happen in just that one place.)

**Review by billsacks — CHANGES_REQUESTED** [review]
Thanks a lot for your work on this @sunnivin ! I have a number of comments inline. I'd be happy to schedule a time to go over this by Zoom if that is easier.

As a reminder, the checkbox / todo workflow & tooling is documented here: https://github.com/ESMCI/github-tools/wiki/Recommended-pull-request-workflow-with-gh-pr-query

**Comment by billsacks** [comment]
- [ ] One more request: @sunnivin or @ ziu1986 can one of you please open an issue noting the need to figure out what to do about the sunlit vs. shaded distinction? i.e., right now we calculate both sunlit and shaded stress coefficients, but only use the sunlit coefficient in LUNA. I can't remember the details of what we discussed about this, but I think that, in conversations with you and @danicalombardozzi , we talked about opening an issue to come back to this.

**Comment by billsacks** [comment]
> can one of you please open an issue noting the need to figure out what to do about the sunlit vs. shaded distinction?

- [ ] Along with an issue, it would also be helpful to call attention to this in the relevant part of the code. Probably the best place to do this would be in the LUNA routine that uses the stress coefficient. Once you have opened an issue, I like referencing the issue in the code with this format:

``` Fortran
  ! BUG(wjs, 2021-05-19, ESCOMP/CTSM#xxx)
```

**Comment by sunnivin** [comment]
Awesome @billsacks ! 

I'm on the list!

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (ziu1986) [reply bld/namelist_files/namelist_definition_ctsm.xml:102]: I see your point and consistency would be great, @billsacks. 
But I do not have any strong opinion on whether it should be "ozone" or "o3". Are there any guidelines regarding the usage of chemical formula vs colloquial names of species?

File: src/biogeophys/OzoneMod.F90
Comment (ziu1986) [reply src/biogeophys/OzoneMod.F90:128]: @billsacks For the time being, we could use the broadleaf parameters for the other PFTs as well. Albeit, the "damage" on needleleaf trees would probably be too high compared with observations.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (ziu1986) [reply bld/namelist_files/namelist_definition_ctsm.xml:90]: @sunnivin and I had a chat regarding this. Referring to what we discussed in our zoom call a while ago, all references to o3_ppbv and o3_type should be removed from this PR. It was probably unintentionally that they ended up in this PR. Our plan is to include these in another PR to separate physics implementation from pure technical handy work.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (billsacks) [reply bld/namelist_files/namelist_definition_ctsm.xml:102]: At least for co2 and ch4, we use the chemical names. So I guess this implies using o3_method instead of ozone_method. I don't have strong feelings myself, though. As I write this out, though, I think I prefer something like o3_stress_method or o3_veg_stress_method rather than just o3_method: I'm realizing that o3_method is pretty vague. (Sorry, I think I might have suggested that name earlier.)

File: src/biogeophys/OzoneMod.F90
Comment (billsacks) [reply src/biogeophys/OzoneMod.F90:128]: I'll leave it up to your judgement.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (sunnivin) [reply bld/namelist_files/namelist_definition_ctsm.xml:90]: Actually, I left this one because I was not 100% sure about what we wanted to do. I think @billsacks told us that somebody is/there are plans for working on the atmospheric forcing, but I think @ziu1986 is interested in having the o3_ppbv option for her scientific explorations. 

Anyway, I totally agree that it should not be included in this PR.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (sunnivin) [reply bld/namelist_files/namelist_definition_ctsm.xml]: Great comments about the naming convention! 

Would it make sense to do: 
```
ozone_metod --> o3_veg_stress_method 
o3_type --> o3_veg_stress_type 
o3_ppbv --> o3_veg_stress_ppbv 
```

Or does this get a little overkill? Personally, I also prefer to be precise rather than vague.

File: src/biogeophys/OzoneMod.F90
Comment (sunnivin) [reply src/biogeophys/OzoneMod.F90:693]: Thank you for elaborating!

File: src/biogeophys/OzoneMod.F90
Comment (sunnivin) [reply src/biogeophys/OzoneMod.F90:66]: I'm a bit confused about the difference between using a ReadNamelist subroutine in the type or not. 

I wrote a suggestion, looking forward to learning this properly in our zoom-call.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (billsacks) [reply bld/namelist_files/namelist_definition_ctsm.xml]: The o3 forcing may be used for other applications besides just this veg stress parameterization. So o3_type and o3_ppbv probably shouldn't have "veg_stress" in their names. I could see renaming o3_type to o3_forcing_type. But it sounds like we're going to remove that variable from this PR, so we can come back to that naming.

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (billsacks) [reply bld/namelist_files/namelist_definition_ctsm.xml:1099]: - [ ] Sorry, I should have pointed to an example earlier: this should explicitly give the different options. An example you can follow is `snow_cover_fraction_method`. It's important that the available options be given in the description, because this appears in auto-generated web documentation. (I don't think the "valid_values" are used in the auto-generation.)

File: src/main/controlMod.F90
Comment (billsacks) [reply src/main/controlMod.F90]: - [ ] Thanks for fixing the type. However, the size (second argument) also needs to be changed: You need to be sure to broadcast the full size of this variable. (As written, I think this will only broadcast the first character.) Look at some of the character variables near the top of this subroutine for examples.

(We should really rewrite this subroutine to use shr_mpi_bcast, which hides these details.)

Comment (sunnivin) [reply src/main/controlMod.F90]: That was embarrassingly sloppy of me. Thank you for pointing this out (again)!

File: src/main/controlMod.F90
Comment (billsacks) [inline src/main/controlMod.F90:455]: - [ ] oops

File: src/biogeophys/OzoneMod.F90
Comment (ekluzek) [reply src/biogeophys/OzoneMod.F90:66]: If you need to read in from a namelist for a type, it's good to put it in a single subroutine that's private to that module (if called from the Init method for the type). Right now it looks like there isn't a namelist read in for this module, so it's unneeded now. But, it sounds like you eventually are going to want a namelist in order to have a toggle to determine what type of ozone stress is going to be used (right now it's hardcoded to one option for "stess_method").

But, I would say you add the ReadNamelist private subroutine, only when it's needed to set that toggle. We envision using namelists to set logical toggles to set different control options in the model (so for things like stress_method"). So I agree with Bill that you should remove this until you are using it to set stress_method.

Hopefully, that helps some...

File: src/main/controlMod.F90
Comment (sunnivin) [reply src/main/controlMod.F90]: Woops! Never push without tests!

File: src/biogeophys/OzoneMod.F90
Comment (sunnivin) [reply src/biogeophys/OzoneMod.F90]: I have now tried to add the parameter (ozone_method --> o3_veg_stress_method) in the input parameters. 

Running the following simple tests 

    SMS_Ld1_P4x1.f10_f10_musgs.I2000Clm50BgcCropQianRs.sunnivin_gnu.clm-o3 
    SMS_Ld1_P4x1.f10_f10_musgs.I2000Clm50BgcCropQianRs.sunnivin_gnu.clm-irrig_o3_reduceOutput
    SMS_Ld1_P4x1.f10_f10_musgs.I2000Clm50BgcCropQianRs.sunnivin_gnu.clm-default 

above they are passed without error. For me it seems logic to also update `param_inst`, but I'm still not sure if this is done correctly now?

**Comment by sunnivin** [comment]
### I'm wondering if it would make sense to update the tests for `o3` a bit? 

To catch a test with `o3_veg_stress_method=stress_falk` would it make sense to restructure the tests in the folder `CTSMROOT/cime_config/testdefs/testmods_dirs/clm/x`, `x=[o3,irrig_o3_reduceOutput]`? 

For me it seems natural to follow the structure by for instance the `CTSMROOT/cime_config/testdefs/testmods_dirs/clm/clmMEC`-test? 

I'm thinking that we could name the new test folders like  
`CTSMROOT/cime_config/testdefs/testmods_dirs/clm/x_y, x=[o3,irrig_o3_reduceOutput], y=[lombardozzi2015, stress_falk]`?

**Comment by sunnivin** [comment]
### Do you prefer to rebase or merge? 

My development branch is now quite far behind ctsm's main. Do you prefer that I do a rebase or merge? 

Furthermore my commit history is not exactly pretty at the current stage. I can take the time to clean up a bit if you like?

**Comment by ekluzek** [comment]
@sunnivin I've personally usually done merges rather than rebase on PR branches. The end result should get you to the same point. The main reason I would see to prefer one over the other is if one has fewer conflicts than the other and as such is easier to do. So I would leave it up to you to decide. @billsacks is there a reason to prefer one over the other?

I would say "having a messy commit history" is a feature rather than a bug. I'd rather see the smaller commits, and have the history of what people did. It tends to be helpful to have smaller commits out there (for example when using git bisect), or just examining when things changed. Mostly that tends to be helpful for when I'm looking at my own commits. But, I have no problem with seeing others as well. If we had a huge developer base it's possible that smaller commits would be a problem, but I've never seen it that way.

**Comment by billsacks** [comment]
Regarding rebase vs. merge: I generally prefer merges for updating a branch. The main times I use rebase is if I recently started some work then realized I should have started from a different point, or if I'm completely moving the branch history from one place to another. The reason I prefer merge is that it preserves the history of what happens rather than rewriting history. This is important for a few reasons, such as: (1) if you documented testing that was done on some hash; (2) if you have already opened a PR and people have reviewed it (I, for one, often make note of which commits I have already reviewed, so if the author rebases their branch it can throw me off); (3) if we need to go back and figure out what was done, e.g., to diagnose a problem that might have come in with the update to master, then using rebase can make this hard or impossible in some scenarios.

Regarding messy commit history: I don't care about this and feel it usually isn't worth the time to go back and clean up. Definitely don't rewrite history of anything you have already pushed to a PR: as with rebasing, that can really mess with the reviewers (at least me). I will sometimes clean up commits that I just made and haven't pushed anywhere yet – mainly in situations where I realized I should have made some other small change to a commit that I recently made, in which case I'll amend the commit or (if I have already made an intervening commit) will do a small interactive rebase to reorder and squash a few commits before pushing my branch.

I'll reply to your other questions soon.

**Review by billsacks — CHANGES_REQUESTED** [review]
Thanks for these changes. They mostly look good, but see some very minor comments below (4 out of 5 of them are just the same spelling mistake repeated in a few places).

File: src/biogeophys/OzoneMod.F90
Comment (billsacks) [reply src/biogeophys/OzoneMod.F90]: It looks like the addition of this new parameter is done mostly right, based on the code anyway (except see my comment about a minor change to namelist_definition).

> For me it seems logic to also update param_inst, but I'm still not sure if this is done correctly now?

Sorry, I'm not sure what you mean by this.

**Comment by billsacks** [comment]
> ### I'm wondering if it would make sense to update the tests for `o3` a bit?
> To catch a test with `o3_veg_stress_method=stress_falk` would it make sense to restructure the tests in the folder `CTSMROOT/cime_config/testdefs/testmods_dirs/clm/x`, `x=[o3,irrig_o3_reduceOutput]`?
> 
> For me it seems natural to follow the structure by for instance the `CTSMROOT/cime_config/testdefs/testmods_dirs/clm/clmMEC`-test?
> 
> I'm thinking that we could name the new test folders like
> `CTSMROOT/cime_config/testdefs/testmods_dirs/clm/x_y, x=[o3,irrig_o3_reduceOutput], y=[lombardozzi2015, stress_falk]`?

I remember going back and forth last time we talked as to whether it was worth changing or adding a test to cover this new code. But now I'll say yes:

- [ ] Let's change one test to cover this new code

I think it's sufficient to have a total of two ozone tests, but let's change one of the existing tests to the Falk method. So we could have testmods named o3lombardozzi2015 (and change the `ERP_D_Ld5_P48x1.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-o3` test to use that testmod) and irrig_o3falk_reduceOutput (and change the `ERP_P72x2_Ly3.f10_f10_mg37.I2000Clm50BgcCrop.cheyenne_intel.clm-irrig_o3_reduceOutput` test to use that testmod) (or the other way around).

- [ ] When doing final testing, though, we should recreate testmod directories corresponding to what's in place now and run tests equivalent to what are currently in the test suite. This is important so that we can do baseline comparisons of both of these tests.

- [ ] In addition, as we discussed a few weeks ago, I'd suggest running one test (as a one-time thing) that turns on the Falk method and is a global `ERP_PNx2_D` test (for some value of N = number of MPI tasks). It's helpful to run a test like this covering new code.

Finally, I had two other items on my list from when we talked last.

- [ ] You were going to add an error check: can only use the Falk ozone method if LUNA is on

- [ ] You were going to add an end-of-day check as noted in https://github.com/ESCOMP/CTSM/pull/1370#discussion_r635534814 (though that isn't absolutely necessary if you're out of time)

File: bld/namelist_files/namelist_definition_ctsm.xml
Comment (sunnivin) [reply bld/namelist_files/namelist_definition_ctsm.xml:1103]: 😅

File: src/biogeophys/OzoneMod.F90
Comment (sunnivin) [reply src/biogeophys/OzoneMod.F90]: > It looks like the addition of this new parameter is done mostly right, based on the code anyway (except see my comment about a minor change to namelist_definition).
> 
> > For me it seems logic to also update param_inst, but I'm still not sure if this is done correctly now?
> 
> Sorry, I'm not sure what you mean by this.

Sorry for expressing myself unclearly, but you answered my question anyway.

File: src/biogeophys/OzoneMod.F90
Comment (sunnivin) [reply src/biogeophys/OzoneMod.F90:659]: Thank you for the remainder of this @billsacks, I do have time for adding this change as well. 

To be 100% sure that I have understood correctly: 
- We introduce a public function called `is_time_to_run_luna` in `src/biogeophys/LunaMod.F90'. This is essentially a wrapper around the `is_end_curr_day`-function. 
- Inside `LunaMod.F90` we do the change `is_end_curr_day --> is_time_to_run_luna` 
- Inside `src/biogeophys/OzoneMod.F90` and the subroutine `CalcOzoneStressFalk` I introduce a conditional check so the current function body is only executed when `is_time_to_run_luna` is evaluated to true.

**Comment by sunnivin** [comment]
> I think it's sufficient to have a total of two ozone tests, but let's change one of the existing tests to the Falk method. So we could have testmods named o3lombardozzi2015 (and change the `ERP_D_Ld5_P48x1.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-o3` test to use that testmod) and irrig_o3falk_reduceOutput (and change the `ERP_P72x2_Ly3.f10_f10_mg37.I2000Clm50BgcCrop.cheyenne_intel.clm-irrig_o3_reduceOutput` test to use that testmod) (or the other way around).
> 

This is clear!

**Comment by sunnivin** [comment]
> * [ ]  When doing final testing, though, we should recreate testmod directories corresponding to what's in place now and run tests equivalent to what are currently in the test suite. This is important so that we can do baseline comparisons of both of these tests.
> * [ ]  In addition, as we discussed a few weeks ago, I'd suggest running one test (as a one-time thing) that turns on the Falk method and is a global `ERP_PNx2_D` test (for some value of N = number of MPI tasks). It's helpful to run a test like this covering new code.

I think that what you mean here is that I can update the file `cime_config/testdefs/testlist_clm.xml` with the changed name of the testdirectories, but when the final testing is done for the PR we have to revert the names back so the basemod comparisons are like before? 

It is indeed a good idea to run the global test case for `o3_veg_stress_method=stress_falk`. I can do that on our cluster when I have merged in the latest version of the current master branch.

File: src/biogeophys/OzoneMod.F90
Comment (billsacks) [reply src/biogeophys/OzoneMod.F90:659]: Yes, that sounds right - thank you!

**Comment by billsacks** [comment]
> I think that what you mean here is that I can update the file `cime_config/testdefs/testlist_clm.xml` with the changed name of the testdirectories, but when the final testing is done for the PR we have to revert the names back so the basemod comparisons are like before?

Yes. For final testing we want to run both the new tests and the two original tests. We can do that by temporarily putting back in place testmod directories with the old names (which both use the Lombardozzi method) (no need to commit them to the repository) then running tests using those test mods. I can help run those if what I'm saying isn't clear – it's easier to do than to explain :-)

**Review by billsacks — APPROVED** [review]
Looks great now - thank you, @sunnivin ! I'm happy to do final testing on this, but I'm thinking I'll do it at a time when I can screen share with @ziu1986 and anyone else who's interested, to walk them through my process.

**Comment by billsacks** [comment]
Initial testing went pretty smoothly, thanks to all of the work put in by @sunnivin and @ziu1986 . There was one subtle restart issue that just impacted the new ozone diagnostic (history) fields: these were only being set in the exposed veg filter, which means that their values were indeterminate in patches that had been exposed earlier in the simulation but were now snow-covered. I have fixed that by explicitly resetting values to 1 every time step in the noexposedveg filter. See https://github.com/billsacks/ctsm/commit/afad47030aa0bef3ce3f8ce194403ee7a2dcefeb

This ties in with a question that came up in our discussion a few months ago: should these non-exposed veg patches contribute to grid cell averages? I have expanded on the comment in the ozone initHistory routine with some thoughts on this that we'll want to return to later.

I am now rerunning full testing.

**Comment by billsacks** [comment]
One other minor issue was that pgi tests changed answers with ozone off (other compilers did not). I have verified that this comes from the extra term in LUNA: multiplying by 1 must lead to roundoff-level changes. I have verified that, if I introduce this diff off of master:

https://github.com/billsacks/ctsm/commit/011c79fc361a98c04a42824b74ca81209262da7c

it is bit-for-bit with this branch.

**Comment by ziu1986** [comment]
> Initial testing went pretty smoothly, thanks to all of the work put in by @sunnivin and @ziu1986 . There was one subtle restart issue that just impacted the new ozone diagnostic (history) fields: these were only being set in the exposed veg filter, which means that their values were indeterminate in patches that had been exposed earlier in the simulation but were now snow-covered. I have fixed that by explicitly resetting values to 1 every time step in the noexposedveg filter. See [billsacks@afad470](https://github.com/billsacks/ctsm/commit/afad47030aa0bef3ce3f8ce194403ee7a2dcefeb)

> I am now rerunning full testing.

Thanks a lot @billsacks and @sunnivin for making this PR come true!

> 
> This ties in with a question that came up in our discussion a few months ago: should these non-exposed veg patches contribute to grid cell averages? I have expanded on the comment in the ozone initHistory routine with some thoughts on this that we'll want to return to later.
> 
I remember that I came across this issue some years ago when adapting code in another model. Back then (if I remember correctly) I decided on dropping non-expo fields from the average.

I have mixed feelings about it now. Because it depends on the use case of the field: 
a.) Coupling to the atmosphere (I suppose still happens at grid level) expects the "real" grid average including the non-expo patches. Else ozone would be biased.
b.) Diagnostics of plant exposure would benefit from excluding non-expo patches as including them would likely lead to an biased of ozone at grid level. 
The direction of the bias depends on the non-expo patch of cause. Usually, ozone is lower over water bodies while higher over urban canopies or ice/snow covered surfaces.

To cut it short: We might need both eventually.

**Comment by ziu1986** [comment]
> One other minor issue was that pgi tests changed answers with ozone off (other compilers did not). I have verified that this comes from the extra term in LUNA: multiplying by 1 must lead to roundoff-level changes. I have verified that, if I introduce this diff off of master:
> 
> [billsacks@011c79f](https://github.com/billsacks/ctsm/commit/011c79fc361a98c04a42824b74ca81209262da7c)
> 
> it is bit-for-bit with this branch.

I'm curious about the bfb test. Is suspect that the float variable which is multiplied is not exactly 1 numerically, as you said @billsacks. How do you usually handle this?

**Comment by billsacks** [comment]
> I'm curious about the bfb test. I suspect that the float variable which is multiplied is not exactly 1 numerically, as you said @billsacks. How do you usually handle this?

It's exactly 1 (as you can see in https://github.com/billsacks/ctsm/commit/011c79fc361a98c04a42824b74ca81209262da7c – and noting that 1 can be represented exactly in floating point), but the issue (I'm pretty sure) is that introducing this extra term leads the compiler to do something different in terms of order of operations, which is mathematically equivalent (assuming no major compiler bugs) but results in different answers due to floating point math.

I usually handle this similarly to what I did here: coming up with a hypothesis about what is causing the answer changes, then making a minimal set of changes off of master to get the same answer-changing behavior and verifying that the full branch is bit-for-bit with this one-off branch. To state this differently, my goal (which is sometimes easier and sometimes harder to achieve) is the following: If master is point A and the full branch is point C, then I want a point B such that
- C is bit-for-bit with B
- B can be seen from a quick review to be effectively the same as A – i.e., it is obvious from examining the diffs in B that there are no real answer changes other than what might arise from floating point roundoff issues.

Then I can put those two pieces of information together to be confident that C is effectively the same as A. (Or sometimes I'll do it the other way around, so that B is a small one-off from C, and B is bit-for-bit with A.)

I then try to document what I did in the ChangeLog so that I or others can review / audit this verification later if needed.

**Comment by ziu1986** [comment]
> > I'm curious about the bfb test. I suspect that the float variable which is multiplied is not exactly 1 numerically, as you said @billsacks. How do you usually handle this?
> 
> It's exactly 1 (as you can see in [billsacks@011c79f](https://github.com/billsacks/ctsm/commit/011c79fc361a98c04a42824b74ca81209262da7c) – and noting that 1 can be represented exactly in floating point), 

Of course! my bad. Thanks for reminding me :+1: 
>
>but the issue (I'm pretty sure) is that introducing this extra term leads the compiler to do something different in terms of order of operations, which is mathematically equivalent (assuming no major compiler bugs) but results in different answers due to floating point math.

That's what I actually had in mind. I just cannot express it as eloquently (and proper) as you.
> 
> I usually handle this similarly to what I did here: coming up with a hypothesis about what is causing the answer changes, then making a minimal set of changes off of master to get the same answer-changing behavior and verifying that the full branch is bit-for-bit with this one-off branch. To state this differently, my goal (which is sometimes easier and sometimes harder to achieve) is the following: If master is point A and the full branch is point C, then I want a point B such that
> 
>     * C is bit-for-bit with B
> 
>     * B can be seen from a quick review to be effectively the same as A – i.e., it is obvious from examining the diffs in B that there are no real answer changes other than what might arise from floating point roundoff issues.
> 
> 
> Then I can put those two pieces of information together to be confident that C is effectively the same as A. (Or sometimes I'll do it the other way around, so that B is a small one-off from C, and B is bit-for-bit with A.)
> 
> I then try to document what I did in the ChangeLog so that I or others can review / audit this verification later if needed.

Thank you for sharing your insight! :bulb: So, it is not much different from how I usually test results.
It probably takes a good deal of experience to quickly locate these issues.

**Pull request merged**
